---
title: "Arabidposis thaliana amino acid Pathway genes data from AraCyc"
author: "Kevin Bird"
date: "June 7, 2016"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
## Load biomaRt pacakge:


```{r}
library("biomaRt", lib.loc="/usr/lib64/R/library")
```


## Set database source in biomaRt to plant ensembl and select the arabidopsis gene annotation dataset:

```{r}
mart <-useMart(biomart="plants_mart",host="plants.ensembl.org",dataset="athaliana_eg_gene")
```

## Find relevant pathway genes:


1. Track down pathways for biosynthesis and degradtion of amino acids in Arabidopsis using AraCyc:
http://pmn.plantcyc.org/ARA/class-tree?object=Pathways

2. Copy genes from pathways into a csv file

3. Read csv into dataframe, pulling out transcript ID for future biomaRt filtering.
+e.g 
  For biosynthesis:

  
```{r}
  At_aa_Biosynthesis_FullDataFrame <-read.csv("/home/biows3/At_aa_biosynthesis_genes_AraCyc_edit.csv",header=TRUE)

 At_Biosyn_Genes_AraCyc <- as.matrix(At_aa_Biosynthesis_FullDataFrame$Gene.Accession)
```

 
 For catabolism:


```{r}
 At_aa_Catabolism_FullDataFrame <- read.csv("/home/biows3/At_aa_catabolism_genes_AraCyc_edit.csv",header=TRUE)

 At_Catab_Genes_AraCyc <- as.matrix(At_aa_Catabolism_FullDataFrame$Gene.Accession)
```


## Pull pathway gene start and stop positions (and other information if desired) from biomaRt based on transcript ID:


For biosynthesis:


```{r}
At_aa_Biosynthesis_BiomaRt_DataSet <-getBM(attributes=c("chromosome_name","start_position", "end_position","ensembl_gene_id"),filters =c("ensembl_transcript_id"), values=c(At_Biosyn_Genes_AraCyc),mart=mart)
```


For catabolism:


```{r}
At_aa_Catab_BiomaRt_DataSet <-getBM(attributes=c("chromosome_name","start_position", "end_position","ensembl_gene_id"),filters =c("ensembl_transcript_id"), values=c( At_Catab_Genes_AraCyc),mart=mart)
```


## Output biomaRt data sets as tab delimited files.
For biosynthesis


```{r}
write.table(At_aa_Biosynthesis_BiomaRt_DataSet, "/home/biows3/At_aa_Biosynthesis_bioMart_DataSet.txt", sep="\t")
```


For catabolism


```{r}
write.table(At_aa_Catab_BiomaRt_DataSet, "/home/biows3/At_aa_Catab_BiomaRt_DataSet.txt", sep ="\t")
```
