---
title: "Make_AAS"
author: "Kevin Bird"
date: "February 21, 2018"
output: html_document
---
Make GAPIT .GM files from AA gene list
```{r}
#install.packages(foreach)
library(foreach)
library(plyr)
library(readr)
SubsetGM1<-c(NA)
SubsetGM2<-c(NA)
SubsetGM3<-c(NA)
SubsetGM4<-c(NA)

AAGeneListTAIR10<-unique(At_aa_bioMart_DataSet)

#Subset full SNP list based on +- 5kb start and stop positions from BiomaRt output
AASubsetGM1<- foreach(i=1:nrow(AAGeneListTAIR10), .combine='rbind') %do% subset(IMPORTANT_Call_Method_75_GM1, Chromosome == AAGeneListTAIR10$chromosome_name[i] & Positions >= AAGeneListTAIR10$start_position[i]-5000 & Positions <= AAGeneListTAIR10$end_position[i]+5000)

AASubsetGM2<- foreach(i=1:nrow(AAGeneListTAIR10), .combine='rbind') %do% subset(IMPORTANT_Call_Method_75_GM2, Chromsome == AAGeneListTAIR10$chromosome_name[i] & Position >= AAGeneListTAIR10$start_position[i]-5000 & Position <= AAGeneListTAIR10$end_position[i]+5000)

AASubsetGM3<- foreach(i=1:nrow(AAGeneListTAIR10), .combine='rbind') %do% subset(IMPORTANT_Call_Method_75_GM3, Chromsome == AAGeneListTAIR10$chromosome_name[i] & Position >= AAGeneListTAIR10$start_position[i]-5000 & Position <= AAGeneListTAIR10$end_position[i]+5000)

AASubsetGM4<- foreach(i=1:nrow(AAGeneListTAIR10), .combine='rbind') %do% subset(IMPORTANT_Call_Method_75_GM4, Chromsome == AAGeneListTAIR10$chromosome_name[i] & Position >= AAGeneListTAIR10$start_position[i]-5000 & Position <= AAGeneListTAIR10$end_position[i]+5000)

AASubsetGM1=unique(AASubsetGM1)
AASubsetGM2=unique(AASubsetGM2)
AASubsetGM3=unique(AASubsetGM3)
AASubsetGM4=unique(AASubsetGM4)


```

Generate GAPIT .GD files using .GM files as index and name first column "Ecotype_ID"
```{r}
AASubsetGenoGD1<-  GenoFile1[,c(AASubsetGM1$SNP_ID)]
AASubsetGenoGD2<- GenoFile2[,c(AASubsetGM2$Name)]
AASubsetGenoGD3<- GenoFile3[,c(AASubsetGM3$Name)]
AASubsetGenoGD4<- GenoFile4[,c(AASubsetGM4$Name)]

AASubsetGenoGD1<-cbind(GenoFile1$Ecotype_ID,AASubsetGenoGD1)
colnames(AASubsetGenoGD1)[1]<-'Ecotype_ID'
AASubsetGenoGD2<-cbind(GenoFile1$Ecotype_ID,AASubsetGenoGD2)
colnames(AASubsetGenoGD2)[1]<-'Ecotype_ID'
AASubsetGenoGD3<-cbind(GenoFile1$Ecotype_ID,AASubsetGenoGD3)
colnames(AASubsetGenoGD3)[1]<-'Ecotype_ID'
AASubsetGenoGD4<-cbind(GenoFile1$Ecotype_ID,AASubsetGenoGD4)
colnames(AASubsetGenoGD4)[1]<-'Ecotype_ID'

```

Output GD and GM files
```{r}
write_delim(AASubsetGenoGD1,"C:\\Users/Kevin/Documents/At-GenomicSelection-Ruthie/AASubsetGeno/AASubsetGenoGD1.txt",delim = "\t")
write_delim(AASubsetGenoGD2,"C:\\Users/Kevin/Documents/At-GenomicSelection-Ruthie/AASubsetGeno/AASubsetGenoGD2.txt",delim = "\t")
write_delim(AASubsetGenoGD3,"C:\\Users/Kevin/Documents/At-GenomicSelection-Ruthie/AASubsetGeno/AASubsetGenoGD3.txt",delim = "\t")
write_delim(AASubsetGenoGD4,"C:\\Users/Kevin/Documents/At-GenomicSelection-Ruthie/AASubsetGeno/AAsubsetGenoGD4.txt",delim = "\t")

write_delim(AASubsetGM1,"C:\\Users/Kevin/Documents/At-GenomicSelection-Ruthie/AASubsetGeno/AASubsetGenoGM1.txt",delim = "\t")
write_delim(AASubsetGM2,"C:\\Users/Kevin/Documents/At-GenomicSelection-Ruthie/AASubsetGeno/AASubsetGenoGM2.txt",delim = "\t")
write_delim(AASubsetGM3,"C:\\Users/Kevin/Documents/At-GenomicSelection-Ruthie/AASubsetGeno/AASubsetGenoGM3.txt",delim = "\t")
write_delim(AASubsetGM4,"C:\\Users/Kevin/Documents/At-GenomicSelection-Ruthie/AASubsetGeno/AASubsetGenoGM4.txt",delim = "\t")
```

