---
title: "MakeGS"
author: "Kevin Bird"
date: "February 21, 2018"
output: html_document
---
Make GAPIT .GM files from TAIR Gene list
```{r}
library(foreach)
GenicSubsetGM1<-c(NA)
GenicSubsetGM2<-c(NA)
GenicSubsetGM3<-c(NA)
GenicSubsetGM4<-c(NA)

GenicSubsetGM1<- foreach(i=1:nrow(Genic_DataSet), .combine='rbind') %do% subset(IMPORTANT_Call_Method_75_GM1, Chromosome == Genic_DataSet$chromosome_name[i] & Positions >= Genic_DataSet$start_position[i]-5000 & Positions <= Genic_DataSet$end_position[i]+5000)

GenicSubsetGM2<- foreach(i=1:nrow(Genic_DataSet), .combine='rbind') %do% subset(IMPORTANT_Call_Method_75_GM2, Chromsome == Genic_DataSet$chromosome_name[i] & Position >= Genic_DataSet$start_position[i]-5000 & Position <= Genic_DataSet$end_position[i]+5000)

GenicSubsetGM3<- foreach(i=1:nrow(Genic_DataSet), .combine='rbind') %do% subset(IMPORTANT_Call_Method_75_GM3, Chromsome == Genic_DataSet$chromosome_name[i] & Position >= Genic_DataSet$start_position[i]-5000 & Position <= Genic_DataSet$end_position[i]+5000)

GenicSubsetGM4<- foreach(i=1:nrow(Genic_DataSet), .combine='rbind') %do% subset(IMPORTANT_Call_Method_75_GM4, Chromsome == Genic_DataSet$chromosome_name[i] & Position >= Genic_DataSet$start_position[i]-5000 & Position <= Genic_DataSet$end_position[i]+5000)

GenicSubsetGM1=unique(GenicSubsetGM1)
GenicSubsetGM2=unique(GenicSubsetGM2)
GenicSubsetGM3=unique(GenicSubsetGM3)
GenicSubsetGM4=unique(GenicSubsetGM4)


```

Generate GAPIT .GD files using .GM files as index and name first column "Ecotype_ID"
```{r}
GenicSubsetGenoGD1<-  GenoFile1[,c(GenicSubsetGM1$SNP_ID)]
GenicSubsetGenoGD2<- GenoFile2[,c(GenicSubsetGM2$Name)]
GenicSubsetGenoGD3<- GenoFile3[,c(GenicSubsetGM3$Name)]
GenicSubsetGenoGD4<- GenoFile4[,c(GenicSubsetGM4$Name)]

GenicSubsetGenoGD1<-cbind(GenoFile1$Ecotype_ID,GenicSubsetGenoGD1)
colnames(GenicSubsetGenoGD1)[1]<-'Ecotype_ID'
GenicSubsetGenoGD2<-cbind(GenoFile1$Ecotype_ID,GenicSubsetGenoGD2)
colnames(GenicSubsetGenoGD2)[1]<-'Ecotype_ID'
GenicSubsetGenoGD3<-cbind(GenoFile1$Ecotype_ID,GenicSubsetGenoGD3)
colnames(GenicSubsetGenoGD3)[1]<-'Ecotype_ID'
GenicSubsetGenoGD4<-cbind(GenoFile1$Ecotype_ID,GenicSubsetGenoGD4)
colnames(GenicSubsetGenoGD4)[1]<-'Ecotype_ID'

```

Output GD and GM files
```{r}
write_delim(GenicSubsetGenoGD1,"C:\\Users/Kevin/Documents/At-GenomicSelection-Ruthie/GenicGeno//GenicSubsetGenoGD1.txt",delim = "\t")
write_delim(GenicSubsetGenoGD2,"C:\\Users/Kevin/Documents/At-GenomicSelection-Ruthie/GenicGeno/GenicSubsetGenoGD2.txt",delim = "\t")
write_delim(GenicSubsetGenoGD3,"C:\\Users/Kevin/Documents/At-GenomicSelection-Ruthie/GenicGeno/GenicSubsetGenoGD3.txt",delim = "\t")
write_delim(GenicSubsetGenoGD4,"C:\\Users/Kevin/Documents/At-GenomicSelection-Ruthie/GenicGeno/GenicSubsetGenoGD4.txt",delim = "\t")

write_delim(GenicSubsetGM1,"C:\\Users/Kevin/Documents/At-GenomicSelection-Ruthie/GenicGeno/GenicSubsetGM1.txt",delim = "\t")
write_delim(GenicSubsetGM2,"C:\\Users/Kevin/Documents/At-GenomicSelection-Ruthie/GenicGeno/GenicSubsetGM2.txt",delim = "\t")
write_delim(GenicSubsetGM3,"C:\\Users/Kevin/Documents/At-GenomicSelection-Ruthie/GenicGeno/GenicSubsetGM3.txt",delim = "\t")
write_delim(GenicSubsetGM4,"C:\\Users/Kevin/Documents/At-GenomicSelection-Ruthie/GenicGeno/GenicSubsetGM4.txt",delim = "\t")

```
